src/components/navigation/SlideMenuItems.tsx

```diff
@@ -1,86 +1,123 @@
-import React, { useEffect, useMemo, useState } from 'react';
+import React, { useCallback, useEffect, useMemo } from 'react';
 
 import { Collapse, List, ListItem, ListItemIcon, ListItemText } from '@material-ui/core';
 import { ExpandLess, ExpandMore } from '@material-ui/icons';
 import { useHistory } from 'react-router-dom';
 
-import { NavigationItem, UtilityItem } from '../../types';
+import { NavigationItem, UtilityItem } from '@app/types';
 
-import { navigationItemParentMap } from './NavigationItems';
+import { useModal } from '@app/contexts/ui/ModalProvider';
 
-import useStyles from './SlideMenu.jss';
+import { useSlideMenuState } from './SlideMenuContext';
 
-const menuState = {
-  expandedNavItems: new Set<string>(),
-};
+import useStyles from './SlideMenu.jss';
 
 /**
  * List item for the navigation list in the slide menu
  */
 interface NavListItemProps extends NavigationItem {
-  menuClosingAction: () => void;
+  onRequestOpen: () => void;
+  onRequestClose: () => void;
+  isDrawerOpen: boolean;
 }
+
 const NavListItem: React.FC<NavListItemProps> = props => {
-  const [open, setOpen] = useState<boolean>(menuState.expandedNavItems.has(props.id));
+  const {
+    id,
+    subItemList,
+    path,
+    onRequestClose,
+    onRequestOpen,
+    isDrawerOpen,
+    isVisible,
+    icon,
+    text,
+  } = props;
   const history = useHistory();
+  const menuState = useSlideMenuState();
   const classes = useStyles();
-  const hasSubList = !!props.subItemList?.length;
-  const isSelected: boolean = history.location.pathname === props.path;
-
-  // @ATTENTION I'd like to use rrd's useHistory hook - will this cause any issues?
-  useEffect(() => {
-    history.listen(location => {
-      const activeItem = Object.values(navigationItemParentMap).find(
-        item => item.path === location.pathname,
-      );
-      if (activeItem) {
-        activeItem.parents.forEach(parent => menuState.expandedNavItems.add(parent));
-      }
-    });
-  }, [history]);
+  const hasSubList = !!subItemList?.length;
+  const sublistOpen = !!menuState?.expandedNavItems.includes(id);
+  const isSelected: boolean = ((): boolean => {
+    if (menuState?.activeNavItem && menuState?.activeNavItem?.path === path) return true;
+    if (isDrawerOpen) return false;
+    if (menuState?.activeNavItem?.parents.includes(id)) return true;
+    return false;
+  })();
```

Comment:
I appreciate the tendency towards const but this is a bitch and a half to read.  what are your thoughts on the following
```suggestion
  let isSelected = false;
  const activeNav = menuState?.activeNavItem;
  if (activeNav?.path === path) {
    isSelected = true;
  } else if (!isDrawerOpen && activeNav?.parents.includes(id)) {
    isSelected = true;
  }
```

Reply:
I would also appreciate some comments on this if possible, this entire component has already gotten pretty convoluted

Reply:
I personally like the self invoking function for scenarios like these because all of the logic is defined at the same spot you declared your variable and since it's const you know that that's the final value it's gonna have and don't have to worry about it possibly changing somewhere else in the function.

Reply:
I've added some comments to explain the flow https://github.com/resident51/resident51/pull/3/commits/ad9cf714da9c80a0a85a88d5891d7dee55f45ffa